### PR TITLE
Add Gosam

### PR DESCRIPTION
**Pull request by iarspider**
Split from #19394

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-gosam/package.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+class PyGosam(PythonPackage):
+    """The package GoSam allows for the automated calculation of
+       one-loop amplitudes for multi-particle processes in renormalizable
+       quantum field theories."""
+
+    homepage = "https://gosam.hepforge.org"
+    url      = "https://gosam.hepforge.org/downloads/?f=gosam-2.0.4-6d9f1cba.tar.gz"
+
+    version('2.0.4', sha256='faf621c70f66d9dffc16ac5cce66258067f39f686d722a4867eeb759fcde4f44')
+    patch('gosam-1.2.4.patch')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-gosam/package.py]: This patch seems to add Python 3 support, but you restrict things to Python 2.7. Is this patch actually useful then?

Comment (iarspider) [reply var/spack/repos/builtin/packages/py-gosam/package.py]: I was trying to add Python3 support, but the task turned out too complicated. I will remove the patch for now.

Comment (iarspider) [reply var/spack/repos/builtin/packages/py-gosam/package.py]: For now = until the developers make a PY3-compatible version

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-gosam/package.py]: You should remove this line now

File: var/spack/repos/builtin/packages/py-gosam/package.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+       quantum field theories."""
+
+    homepage = "https://gosam.hepforge.org"
+    url      = "https://gosam.hepforge.org/downloads/?f=gosam-2.0.4-6d9f1cba.tar.gz"
+
+    version('2.0.4', sha256='faf621c70f66d9dffc16ac5cce66258067f39f686d722a4867eeb759fcde4f44')
+
+    depends_on('form', type=('build', 'run'))
+    depends_on('qgraf', type=('build', 'run'))
+    depends_on('gosam-contrib', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-gosam/package.py]: In light of your other PRs... are these deptypes correct?

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**